File: re-write-with-typescript/rbac.ts
type ActionEnum =
  | "read"
  | "create"
  | "update"
  | "delete"
  | "*"
  | "!"


type RoleEnum =
  | "admin"
  | "guest"
  | "user"
  | "employee"
  | "manager"


type User = {
  name: string,
  role: RoleEnum
}


class Permission {
  resource: string;
  action: ActionEnum;

  constructor(resource: string, action: ActionEnum) {
    this.resource = resource;
    this.action = action;
  }

  get name() {
    return `${this.resource}:${this.action}`
  }
}


type Role = {
  name: RoleEnum,
  permissions: Permission[]
}


function log_perms(rules: Map<RoleEnum, Role>, role: RoleEnum) {
  const current_rule = rules.get(role);

  if (!current_rule) throw new Error("Not Found Role name");

  for (const perm of current_rule?.permissions) {
    console.log("[ Permission ]", perm);
  }
}


class AccessRule {
  access_rules: Map<RoleEnum, Role> = new Map([
    ["admin", {
      name: "admin", 
      permissions: [
        new Permission("*", "*")
      ]
    }],

    ["guest", { 
      name: "guest", 
      permissions: [
        new Permission("post", "read"),
      ]
    }]
  ]);

  authorize<U extends { role: RoleEnum }>(user: U, permission: string) {
    let access: boolean = false;
    const [resource, action] = permission.split(":");

    log_perms(this.access_rules, user.role);
  
    for (const access_rule of this.access_rules.values()) {
      if (access_rule.name === user.role) {
        for (const permission of access_rule.permissions) {
          if (
            (resource == permission.resource || permission.resource === "*") 
              && (permission.action === action || permission.action === "*")
          ) {
            access = true
          }
          if (resource == permission.resource && permission.action === "!") {
            access = false
          }
        }
      }
    }

    return access;
  }
}


function main() {
  const admin: User = {
    name: "bob",
    role: "admin"
  }

  const anonymous: User = {
    name: "some",
    role: "guest"
  }

  const rbac = new AccessRule();

  const admin_dashboard_edit_perm = rbac.authorize(admin, "dashboard:edit");

  const guest_post_read_perm = rbac.authorize(anonymous, "post:read");
  const guest_dashboard_edit_perm = rbac.authorize(anonymous, "dashboard:edit");

  console.log({ admin_dashboard_edit_perm });
  console.log({ guest_post_read_perm });
  console.log({ guest_dashboard_edit_perm });
}


main()

File: rbac.py
from __future__ import annotations

from dataclasses import dataclass
from enum import Enum
from abc import ABC, abstractmethod
from typing import List
import logging


class Authoritor(ABC):
    """trait authoritor"""
    @abstractmethod
    def is_authorized(self, permission: Permission, role: Role, action: Action) -> bool:
        """check is authorize, return bool"""


class Permission(ABC):
    """trait action creator"""
    @abstractmethod
    def create_allowed_roles(self) -> List[Role]:
        """create read allowed roles, list"""

    @abstractmethod
    def read_allowed_roles(self) -> List[Role]:
        """read read allowed roles, list"""

    @abstractmethod
    def update_allowed_roles(self) -> List[Role]:
        """update read allowed roles, list"""

    @abstractmethod
    def delete_allowed_roles(self) -> List[Role]:
        """delete read allowed roles, list"""


class Role(str, Enum):
    ADMIN = "admin"
    USER = "user"
    GUEST = "guest"
    ALL = "*"


class Action(str, Enum):
    CREATE = "create"
    READ = "read"
    UPDATE = "update"
    DELETE = "delete"


@dataclass(frozen=True)
class User:
    name: str
    role: Role


@dataclass(frozen=True)
class Rbac(Authoritor):
    def is_authorized(self, permission: Permission, role: Role, action: Action) -> bool:
        access = False
        perm: List[Role] = []

        match action:
            case Action.CREATE:
                logging.info("add create")
                perm.extend(permission.create_allowed_roles())
            case Action.READ:
                logging.info("add read")
                perm.extend(permission.read_allowed_roles())
            case Action.UPDATE:
                logging.info("add update")
                perm.extend(permission.update_allowed_roles())
            case Action.DELETE:
                logging.info("add delete")
                perm.extend(permission.delete_allowed_roles())
            case _:
                raise Exception("Unreachable")

        logging.info(f"{perm = }")

        if Role.ALL in perm:
            access = True
        elif role in perm:
            access = True

        return access


@dataclass(frozen=True)
class DashboardPermission(Permission):
    def create_allowed_roles(self) -> List[Role]:
        allowed_roles = [Role.ADMIN, Role.USER]
        return allowed_roles

    def read_allowed_roles(self) -> List[Role]:
        allowed_roles = [Role.ALL]
        return allowed_roles

    def update_allowed_roles(self) -> List[Role]:
        allowed_roles = [Role.ADMIN, Role.USER]
        return allowed_roles

    def delete_allowed_roles(self) -> List[Role]:
        allowed_roles = [Role.ADMIN]
        return allowed_roles


def main():
    logging.basicConfig(
        format="[  ] %(message)s", 
        datefmt="%Y-%m-%d",
        level=logging.WARN
    )
    rbac = Rbac()

    user = User(name="bob", role=Role.GUEST)

    is_allowed = rbac.is_authorized(
        permission=DashboardPermission(), 
        role=user.role, 
        action=Action.UPDATE
    )

    print(is_allowed)


if __name__ == "__main__":
    main()
